crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,150 @@
+    const EVP_MD* evp_md = EVP_get_digestbynid(hash_type);
+    HMAC_CTX* hctx = HMAC_CTX_new();
+
+    const int qlen = BN_num_bits(range);
+    const int rlen = BN_num_bytes(range);
+    const int hlen = EVP_MD_size(evp_md);
+    unsigned char *x = (unsigned char*)OPENSSL_malloc(rlen);
+    unsigned char *h = (unsigned char*)OPENSSL_malloc(rlen);
+    unsigned char V[EVP_MAX_MD_SIZE];
+    unsigned char K[EVP_MAX_MD_SIZE];
+    unsigned char *T = (unsigned char*)OPENSSL_malloc(rlen);
+    if (!int2octects(x, priv, rlen) ||
+        !bits2octects(h, range, message, message_len, ctx))
+        goto err;
+
+    memset(V, 1, hlen);
+    memset(K, 0, hlen);
+
+    if (evp_md == NULL || hctx == NULL)
+    {
+        if (evp_md == NULL)
+            BNerr(BN_F_BN_GENERATE_DSA_DETERMINISTIC_NONCE, BN_R_NO_SUITABLE_DIGEST);
+        goto err;
+    }
+
+    if (!HMAC_CTX_reset(hctx) ||
+        !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+        !HMAC_Update(hctx, V, hlen) ||
+        !HMAC_Update(hctx, constant + 0, 1) ||
+        !HMAC_Update(hctx, x, rlen) ||
+        !HMAC_Update(hctx, h, rlen) ||
+        !HMAC_Final(hctx, K, NULL))
+        goto err;
+
+    if (!HMAC_CTX_reset(hctx) ||
+        !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+        !HMAC_Update(hctx, V, hlen) ||
+        !HMAC_Final(hctx, V, NULL))
+        goto err;
+
+    if (!HMAC_CTX_reset(hctx) ||
+        !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+        !HMAC_Update(hctx, V, hlen) ||
+        !HMAC_Update(hctx, constant + 1, 1) ||
+        !HMAC_Update(hctx, x, rlen) ||
+        !HMAC_Update(hctx, h, rlen) ||
+        !HMAC_Final(hctx, K, NULL))
+        goto err;
+
+    if (!HMAC_CTX_reset(hctx) ||
+        !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+        !HMAC_Update(hctx, V, hlen) ||
+        !HMAC_Final(hctx, V, NULL))
+        goto err;
+
+    while (1)
+    {
+        int offlen = 0;
+        int i;
+        for (i = 0; i < rlen; i += offlen)
+        {
+            if (!HMAC_CTX_reset(hctx) ||
+                !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+                !HMAC_Update(hctx, V, hlen) ||
+                !HMAC_Final(hctx, V, NULL))
+                goto err;
+            offlen = ((rlen - i) < hlen) ? (rlen - i) : hlen;
+            memcpy(T + i, V, offlen);
+        }
+        bits2int(out, qlen, T, rlen);
+        if ((!BN_is_zero(out)) && (!BN_is_one(out)) && (BN_cmp(out, range) < 0))
+            break;
+
+        if (!HMAC_CTX_reset(hctx) ||
+            !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+            !HMAC_Update(hctx, V, hlen) ||
+            !HMAC_Update(hctx, constant + 0, 1) ||
+            !HMAC_Final(hctx, K, NULL))
+            goto err;
+
+        if (!HMAC_CTX_reset(hctx) ||
+            !HMAC_Init_ex(hctx, K, hlen, evp_md, NULL) ||
+            !HMAC_Update(hctx, V, hlen) ||
+            !HMAC_Final(hctx, V, NULL))
+            goto err;
+    }
+    HMAC_CTX_free(hctx);
```

Comment:
instead of doing this - use the pattern
int ret = 0;

    ret = 1;

err:

   cleanup()
   return ret;

Reply:
I've re-written this method, please have a look!